File: leetcode/src/main/java/linear/linkedlist/_13_palindrome_linkedlist_234/Solution2.java
package linear.linkedlist._13_palindrome_linkedlist_234;

import java.util.Deque;
import java.util.LinkedList;
import linear.linkedlist._13_palindrome_linkedlist_234.Main.ListNode;

public class Solution2 {
    public boolean isPalindrome(ListNode head) {
        Deque<Integer> deque = new LinkedList<>();
        // 연결 리스트를 데크에 삽입
        ListNode node = head;
        while (node != null) {
            deque.add(node.val);
            node = node.next;
        }

        // 데크가 모두 비거나(짝수 개일 때) 1개 이하(홀수 개일 때)가 될 때까지 비교
        while (!deque.isEmpty() && deque.size() > 1) {
            // 데크의 양끝을 추출해 팰린드롬 여부 확인
            if (deque.pollFirst() != deque.pollLast()) {
                return false;
            }
        }
        return true;
    }

    // 11~13ms
}

File: leetcode/src/main/java/linear/linkedlist/_14_merge_two_sorted_lists_21/Solution2.kt
package linear.linkedlist._14_merge_two_sorted_lists_21

class Solution2 {
    fun mergeTwoLists(list1: Main.ListNode?, list2: Main.ListNode?): Main.ListNode? {
        // 두 노드 중 한쪽이 널이면 아닌 노드를 리턴
        if (list1 == null) return list2
        if (list2 == null) return list1

        // l2가 더 크면 l1에 재귀 호출 결과를 엮고 l1을 리턴
        return if (list1.`val` < list2.`val`) {
            list1.next = mergeTwoLists(list1.next, list2)
            list1
            // l1이 더 크거나 같으면 l2에 재귀 호출 결과를 엮고 l2를 리턴
        } else {
            list2.next = mergeTwoLists(list1, list2.next)
            list2
        }
    }
}